0004-post-tags.md

```diff
@@ -0,0 +1,299 @@
+- Feature Name: Post Tags
+- Start Date: 2023-11-04
+- RFC PR: [LemmyNet/rfcs#0004](https://github.com/LemmyNet/rfcs/pull/4)
+- Lemmy Issue: [LemmyNet/lemmy#317](https://github.com/LemmyNet/lemmy/issues/317)
+
+# Summary
+
+Ability to add descriptive tags to Posts similar to Reddit's Flair system primarily on a per-community basis.
+
+# Motivation
+
+Post Tags would allow for vastly improved content filtering (and to a lesser extent searching) both for Users and Admins/Moderators. 
+
+# Guide-level explanation
+
+Tags grant you more fine grained control over the content you see on Lemmy. Certain users do not want to see certain content and tags allow you to do exactly that.
+
+For example a lot of communities offer mixed content such as Memes, News and Discussions. Some people would like to only see News or Discussions. In such a case they could blacklist (filter out) the "Meme" tag, going forward they would not be presented with any Memes in that community.
+
+Another use for tags is better handling of how posts are displayed. For example some people would like the content of a post to stay hidden for all users. This could be the case in a community focused on release discussions of Books, Movies, TV Shows or communities with content some users might find offensive but which is not neccesarily NSFW (for example jokes about certain topics).
+
+Essentially the tags serve as an easier way to message what is *inside* the post. Combining multiple tags allows for even finer control. Made a post about news in poland? Add the "News" and "Poland" tags to it and people interested exclusively in polish news can find it easily or people not interested can avoid it just as easily.
+
+# Reference-level explanation
+
+### Protocol:
+According to https://www.w3.org/TR/activitystreams-vocabulary/#dfn-tags a general tag object exists in the ActivityPub protocol. 
```

Comment:
This is not exactly accurate, reading the spec it says that a `tag` property, containing an Object list is part of all Objects. This means that we need to define our own Tag Object type(s). So the tags would look more like this:

```json
{
  "type": "Post",
  "tags": [
    {
      "@type": "https://json-ld.lemmy.ml/CommunityTag",
      "id": "https://lemmy.world/t/FooTag",
      "name": "FooTag"
    }
  ],
  ...
}
```

Reply:
Meaning: The exact properties the tag object should have should probably be part of this RFC. And also specifying if there's multiple different tag types or if they (community+instance tags) have the same interface

Reply:
Just for reference, here's how mastodon uses hashtags:
`curl -H 'Accept: application/activity+json' https://mastodon.social/@nixCraft/111361899987606218`
```
  "tag": [
    {
      "type": "as:Hashtag",
      "href": "https://mastodon.social/tags/linux",
      "name": "#linux"
    }
  ],
```

Using the non-standardized ActivityStreams:HashTag type (https://github.com/w3c/activitypub/issues/235)

Reply:
The tag system for Lemmy seems fundamentally different from hashtags in Mastodon, so I think its reasonable to federate them in a way thats intentionally incompatible.

Reply:
Agree - I just put it as a comparison. We should create our own object type.

Reply:
@Nutomic What is the fundamental difference between tags described in this proposal and hashtags used by microblogging services like Mastodon? I've read the proposal and there are many similarities.

---

btw, the URL in a highlighted section is not correct. It should be https://www.w3.org/TR/activitystreams-vocabulary/#dfn-tag (not `#dfn-tags`)

Reply:
If Lemmy federates a separate custom-built Tag object instead of `Hashtag`s it has the ability to rename tags without having to send a few billion `Update`s to individually edit all posts that have said tags attached to them. (Instead it can just `Update` the individual tag)

This would also allow for a protocol for cross-instance tag moderation to be built, e.g. using `Add/Tag` `Undo/Add/Tag` (`Accept/Add/Tag` and `Reject/Add/Tag` for tag suggestions?) which AFAIK is impossible with the current design of hashtags (and wouldn't federate to existing implementations anyway)

Reply:
> What is the fundamental difference between tags described in this proposal and hashtags used by microblogging services like Mastodon? I've read the proposal and there are many similarities.

Mostly Lemmy tags being static as opposed to dynamically user created. The Lemmy tags exist in a database table and can be mass edited if needed since they will be referenced by id. Plus the general use case for tags on Lemmy seems less a "just randomly tagging this so people can find it elsewhere" and more a "this community has broad interests so I'll specify what's roughly inside this post"

Reply:
> This would also allow for a protocol for cross-instance tag moderation to be built, e.g. using Add/Tag Undo/Add/Tag (Accept/Add/Tag and Reject/Add/Tag for tag suggestions?) which AFAIK is impossible with the current design of hashtags (and wouldn't federate to existing implementations anyway)

Should be possible, changes to tags would still need to be federated but the posts themselves should be able to stay untouched. Simply changing the values in the tag table should "edit" all existing posts automatically.

Reply:
> This is not exactly accurate, reading the spec it says that a tag property, containing an Object list is part of all Objects. This means that we need to define our own Tag Object type(s). So the tags would look more like this:

this seems to be correct, I've added the relevant part to the RFC with b6afe0786f537791cbe7e7cbc20e9245cabac123. I think the initial tag object can be kept simple, it's easier to add things later on than remove them imo.